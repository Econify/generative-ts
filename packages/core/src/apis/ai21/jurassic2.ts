import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/Either";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { EjsTemplate } from "../../utils/ejsTemplate";

import { nullable } from "../_utils/io-ts-nullable";

const templateSource = `{
  "prompt": "<%= prompt %>"
  <% if (typeof numResults !== 'undefined') { %>, "numResults": <%= numResults %><% } %>
  <% if (typeof maxTokens !== 'undefined') { %>, "maxTokens": <%= maxTokens %><% } %>
  <% if (typeof minTokens !== 'undefined') { %>, "minTokens": <%= minTokens %><% } %>
  <% if (typeof temperature !== 'undefined') { %>, "temperature": <%= temperature %><% } %>
  <% if (typeof topP !== 'undefined') { %>, "topP": <%= topP %><% } %>
  <% if (typeof topKReturn !== 'undefined') { %>, "topKReturn": <%= topKReturn %><% } %>  
  <% if (typeof stopSequences !== 'undefined') { %>, "stopSequences": <%- JSON.stringify(stopSequences) %><% } %>
  <% if (typeof countPenalty !== 'undefined') { %>
    , "countPenalty": {
      "scale": <%= countPenalty.scale %>
      <% if (typeof countPenalty.applyToWhitespaces !== 'undefined') { %>
        , "applyToWhitespaces": <%= countPenalty.applyToWhitespaces %>
      <% } %>
      <% if (typeof countPenalty.applyToPunctuations !== 'undefined') { %>
        , "applyToPunctuations": <%= countPenalty.applyToPunctuations %>
      <% } %>
      <% if (typeof countPenalty.applyToNumbers !== 'undefined') { %>
        , "applyToNumbers": <%= countPenalty.applyToNumbers %>
      <% } %>
      <% if (typeof countPenalty.applyToStopwords !== 'undefined') { %>
        , "applyToStopwords": <%= countPenalty.applyToStopwords %>
      <% } %>
      <% if (typeof countPenalty.applyToEmojis !== 'undefined') { %>
        , "applyToEmojis": <%= countPenalty.applyToEmojis %>
      <% } %>
    }
  <% } %>
  <% if (typeof presencePenalty !== 'undefined') { %>
    , "presencePenalty": {
      "scale": <%= presencePenalty.scale %>
      <% if (typeof presencePenalty.applyToWhitespaces !== 'undefined') { %>
        , "applyToWhitespaces": <%= presencePenalty.applyToWhitespaces %>
      <% } %>
      <% if (typeof presencePenalty.applyToPunctuations !== 'undefined') { %>
        , "applyToPunctuations": <%= presencePenalty.applyToPunctuations %>
      <% } %>
      <% if (typeof presencePenalty.applyToNumbers !== 'undefined') { %>
        , "applyToNumbers": <%= presencePenalty.applyToNumbers %>
      <% } %>
      <% if (typeof presencePenalty.applyToStopwords !== 'undefined') { %>
        , "applyToStopwords": <%= presencePenalty.applyToStopwords %>
      <% } %>
      <% if (typeof presencePenalty.applyToEmojis !== 'undefined') { %>
        , "applyToEmojis": <%= presencePenalty.applyToEmojis %>
      <% } %>
    }
  <% } %>
  <% if (typeof frequencyPenalty !== 'undefined') { %>
    , "frequencyPenalty": {
      "scale": <%= frequencyPenalty.scale %>
      <% if (typeof frequencyPenalty.applyToWhitespaces !== 'undefined') { %>
        , "applyToWhitespaces": <%= frequencyPenalty.applyToWhitespaces %>
      <% } %>
      <% if (typeof frequencyPenalty.applyToPunctuations !== 'undefined') { %>
        , "applyToPunctuations": <%= frequencyPenalty.applyToPunctuations %>
      <% } %>
      <% if (typeof frequencyPenalty.applyToNumbers !== 'undefined') { %>
        , "applyToNumbers": <%= frequencyPenalty.applyToNumbers %>
      <% } %>
      <% if (typeof frequencyPenalty.applyToStopwords !== 'undefined') { %>
        , "applyToStopwords": <%= frequencyPenalty.applyToStopwords %>
      <% } %>
      <% if (typeof frequencyPenalty.applyToEmojis !== 'undefined') { %>
        , "applyToEmojis": <%= frequencyPenalty.applyToEmojis %>
      <% } %>
    }
  <% } %>
}`;

interface PenaltyOptions {
  scale: number;
  applyToWhitespaces?: boolean;
  applyToPunctuations?: boolean;
  applyToNumbers?: boolean;
  applyToStopwords?: boolean;
  applyToEmojis?: boolean;
}

/**
 * @category Ai21 Jurassic 2
 * @category Requests
 */
export interface Ai21Jurassic2Options extends ModelRequestOptions {
  numResults?: number;
  maxTokens?: number;
  minTokens?: number;
  temperature?: number;
  topP?: number;
  topKReturn?: number;
  stopSequences?: string[];
  countPenalty?: PenaltyOptions;
  presencePenalty?: PenaltyOptions;
  frequencyPenalty?: PenaltyOptions;
}

/**
 * @category Ai21 Jurassic 2
 * @category Templates
 */
export const Ai21Jurassic2Template = new EjsTemplate<Ai21Jurassic2Options>(
  templateSource,
);

const Ai21Jurassic2ResponseCodec = t.type({
  id: t.number,
  prompt: t.type({
    text: t.string,
    tokens: t.array(
      t.type({
        generatedToken: t.type({
          token: t.string,
          logprob: t.number,
          raw_logprob: t.number,
        }),
        topTokens: nullable(
          t.union([
            t.string,
            t.array(
              t.type({
                token: t.string,
                logprob: t.number,
                raw_logprob: t.number,
              }),
            ),
          ]),
        ),
        textRange: t.type({
          start: t.number,
          end: t.number,
        }),
      }),
    ),
  }),
  completions: t.array(
    t.type({
      data: t.type({
        text: t.string,
        tokens: t.array(
          t.type({
            generatedToken: t.type({
              token: t.string,
              logprob: t.number,
              raw_logprob: t.number,
            }),
            topTokens: nullable(
              t.union([
                t.string,
                t.array(
                  t.type({
                    token: t.string,
                    logprob: t.number,
                    raw_logprob: t.number,
                  }),
                ),
              ]),
            ),
            textRange: t.type({
              start: t.number,
              end: t.number,
            }),
          }),
        ),
      }),
      finishReason: t.intersection([
        t.type({
          reason: t.string,
        }),
        t.partial({
          length: t.number,
        }),
      ]),
    }),
  ),
});

/**
 * @category Ai21 Jurassic 2
 * @category Responses
 */
export interface Ai21Jurassic2Response
  extends TypeOf<typeof Ai21Jurassic2ResponseCodec> {}

export function isAi21Jurassic2Response(
  response: unknown,
): response is Ai21Jurassic2Response {
  return !isLeft(Ai21Jurassic2ResponseCodec.decode(response));
}

export interface Ai21Jurassic2Api
  extends ModelApi<Ai21Jurassic2Options, Ai21Jurassic2Response> {}

/**
 *
 * ## Reference
 * [Ai21 Jurrassic 2](https://docs.ai21.com/reference/j2-complete-ref)
 *
 * ## Providers using this API
 * - {@link createAwsBedrockModelProvider | AWS Bedrock}
 *
 * @category APIs
 * @category Ai21 Jurassic 2
 * @category Provider: AWS Bedrock
 *
 */
export const Ai21Jurassic2Api: Ai21Jurassic2Api = {
  requestTemplate: Ai21Jurassic2Template,
  responseGuard: isAi21Jurassic2Response,
};
