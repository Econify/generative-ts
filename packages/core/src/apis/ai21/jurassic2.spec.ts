import { Ai21Jurassic2Api, Ai21Jurassic2Options } from "./jurassic2";

function render(context: Omit<Ai21Jurassic2Options, "modelId">) {
  try {
    const rendered = Ai21Jurassic2Api.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("Ai21Jurassic2Api:", () => {
  /**
   * FewShotRequestOptions (prompt):
   */

  test("prompt", () => {
    const rendered = render({
      prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      prompt: "mock-prompt",
      numResults: 3,
      maxTokens: 200,
      minTokens: 100,
      temperature: 0.7,
      topP: 0.9,
      topKReturn: 5,
      stopSequences: ["mock-stop-1", "mock-stop-2"],
      countPenalty: {
        scale: 0.5,
        applyToWhitespaces: true,
        applyToPunctuations: false,
        applyToNumbers: true,
        applyToStopwords: false,
        applyToEmojis: true,
      },
      presencePenalty: {
        scale: 0.3,
        applyToWhitespaces: false,
        applyToPunctuations: true,
        applyToNumbers: false,
        applyToStopwords: true,
        applyToEmojis: false,
      },
      frequencyPenalty: {
        scale: 0.7,
        applyToWhitespaces: true,
        applyToPunctuations: true,
        applyToNumbers: false,
        applyToStopwords: false,
        applyToEmojis: true,
      },
    });

    expect(rendered).toMatchSnapshot();
  });
});
