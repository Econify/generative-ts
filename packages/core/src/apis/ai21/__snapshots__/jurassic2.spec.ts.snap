// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ai21Jurassic2Api: all options 1`] = `
{
  "countPenalty": {
    "applyToEmojis": true,
    "applyToNumbers": true,
    "applyToPunctuations": false,
    "applyToStopwords": false,
    "applyToWhitespaces": true,
    "scale": 0.5,
  },
  "frequencyPenalty": {
    "applyToEmojis": true,
    "applyToNumbers": false,
    "applyToPunctuations": true,
    "applyToStopwords": false,
    "applyToWhitespaces": true,
    "scale": 0.7,
  },
  "maxTokens": 200,
  "minTokens": 100,
  "numResults": 3,
  "presencePenalty": {
    "applyToEmojis": false,
    "applyToNumbers": false,
    "applyToPunctuations": true,
    "applyToStopwords": true,
    "applyToWhitespaces": false,
    "scale": 0.3,
  },
  "prompt": "mock-prompt",
  "stopSequences": [
    "mock-stop-1",
    "mock-stop-2",
  ],
  "temperature": 0.7,
  "topKReturn": 5,
  "topP": 0.9,
}
`;

exports[`Ai21Jurassic2Api: prompt 1`] = `
{
  "prompt": "mock-prompt",
}
`;
